sdk/core/azure-core/azure/core/messaging.py

```diff
@@ -192,26 +192,26 @@ def from_dict(cls, event: Dict[str, Any]) -> CloudEvent[DataType]:
         except KeyError as err:
             # https://github.com/cloudevents/spec Cloud event spec requires source, type,
             # specversion. We autopopulate everything other than source, type.
-            if not all(_ in event for _ in ("source", "type")):
-                if all(
-                    _ in event
-                    for _ in (
-                        "subject",
-                        "eventType",
-                        "data",
-                        "dataVersion",
-                        "id",
-                        "eventTime",
-                    )
-                ):
-                    raise ValueError(
-                        "The event you are trying to parse follows the Eventgrid Schema. You can parse"
-                        + " EventGrid events using EventGridEvent.from_dict method in the azure-eventgrid library."
-                    ) from err
+            # So we will assume the KeyError is coming from source/type access.
```

Comment:
This was an actual issue found by pyright:
```
Cell In[8], line 1
----> 1 CloudEvent.from_dict({
      2     "source": 42,
      3     "type": 78,
      4     "time": {"foo": True}   })

File ~/git/azure-sdk-for-python/sdk/core/azure-core/azure/core/messaging.py:215, in CloudEvent.from_dict(cls, event)
    207             raise ValueError(
    208                 "The event you are trying to parse follows the Eventgrid Schema. You can parse"
    209                 + " EventGrid events using EventGridEvent.from_dict method in the azure-eventgrid library."
    210             ) from err
    211         raise ValueError(
    212             "The event does not conform to the cloud event spec https://github.com/cloudevents/spec."
    213             + " The `source` and `type` params are required."
    214         ) from err
--> 215 return event_obj

UnboundLocalError: cannot access local variable 'event_obj' where it is not associated with a value
```

This is replacing it with an actual error that says your dict is invalid, which is way better.

Reply:
https://github.com/Azure/azure-sdk-for-python/issues/31835

Reply:
I'm fine with this fix. Though an alternative would be so simply add `raise` to the end of the except clause so that even if our checks for known structure are clean, the original `KeyError` from the time parsing/elsewhere gets raised (assuming that original error was useful in some way).

Reply:
I feel an error that says "your dict is supposed to be valid CloudEvent" is more helpful on this one than a generic KeyError